File: src/EyeTimeTracker.Core/Sync/UsageSegmentMerger.cs
using EyeTimeTracker.Core.Models;

namespace EyeTimeTracker.Core.Sync;

public static class UsageSegmentMerger
{
    private const long BucketSeconds = 10;
    private const long SessionBreakSeconds = 180;

    public static DailyRecord BuildDailyRecord(DateOnly date, IEnumerable<UsageSegment> segments)
    {
        var buckets = new SortedSet<long>();
        foreach (var segment in segments)
        {
            if (segment.EndUnixSeconds <= segment.StartUnixSeconds)
            {
                continue;
            }

            var startBucket = FloorBucket(segment.StartUnixSeconds);
            var endBucket = CeilBucket(segment.EndUnixSeconds);
            for (var bucket = startBucket; bucket < endBucket; bucket++)
            {
                var bucketDate = DateOnly.FromDateTime(DateTimeOffset.FromUnixTimeSeconds(bucket * BucketSeconds).LocalDateTime);
                if (bucketDate == date)
                {
                    buckets.Add(bucket);
                }
            }
        }

        var record = new DailyRecord(date)
        {
            TotalSeconds = buckets.Count * BucketSeconds,
            HourlySeconds = BuildHourlySeconds(date, buckets),
            SessionSeconds = BuildSessionSeconds(buckets),
            CurrentSessionSeconds = 0
        };
        return record;
    }

    private static long[] BuildHourlySeconds(DateOnly date, IEnumerable<long> buckets)
    {
        var hourly = new long[24];
        foreach (var bucket in buckets)
        {
            var local = DateTimeOffset.FromUnixTimeSeconds(bucket * BucketSeconds).LocalDateTime;
            if (DateOnly.FromDateTime(local) == date)
            {
                hourly[local.Hour] += BucketSeconds;
            }
        }

        return hourly;
    }

    private static List<long> BuildSessionSeconds(IEnumerable<long> buckets)
    {
        var sessions = new List<long>();
        long? previousBucket = null;
        var currentSeconds = 0L;

        foreach (var bucket in buckets)
        {
            if (previousBucket is not null)
            {
                var gapSeconds = (bucket - previousBucket.Value) * BucketSeconds;
                if (gapSeconds > SessionBreakSeconds)
                {
                    if (currentSeconds > 0)
                    {
                        sessions.Add(currentSeconds);
                    }

                    currentSeconds = 0;
                }
            }

            currentSeconds += BucketSeconds;
            previousBucket = bucket;
        }

        if (currentSeconds > 0)
        {
            sessions.Add(currentSeconds);
        }

        return sessions;
    }

    private static long FloorBucket(long unixSeconds)
    {
        return unixSeconds / BucketSeconds;
    }

    private static long CeilBucket(long unixSeconds)
    {
        return (unixSeconds + BucketSeconds - 1) / BucketSeconds;
    }
}

File: android/EyeTimeTrackerAndroid/src/com/eyetimetracker/android/UsageSegmentMerger.java
package com.eyetimetracker.android;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import java.util.SortedSet;
import java.util.TreeSet;

public final class UsageSegmentMerger {
    private static final long BUCKET_SECONDS = 10L;
    private static final long SESSION_BREAK_SECONDS = 180L;

    private UsageSegmentMerger() {
    }

    public static DailySummary buildDailySummary(String date, List<UsageSegment> segments) {
        LocalDate targetDate = LocalDate.parse(date);
        SortedSet<Long> buckets = new TreeSet<>();
        if (segments != null) {
            for (UsageSegment segment : segments) {
                if (segment == null || segment.endUnixSeconds <= segment.startUnixSeconds) {
                    continue;
                }

                long startBucket = floorBucket(segment.startUnixSeconds);
                long endBucket = ceilBucket(segment.endUnixSeconds);
                for (long bucket = startBucket; bucket < endBucket; bucket++) {
                    LocalDate bucketDate = localDateForBucket(bucket);
                    if (targetDate.equals(bucketDate)) {
                        buckets.add(bucket);
                    }
                }
            }
        }

        return new DailySummary(
                date,
                buckets.size() * BUCKET_SECONDS,
                buildHourlySeconds(targetDate, buckets),
                buildSessionSeconds(buckets),
                0L,
                false,
                0);
    }

    private static long[] buildHourlySeconds(LocalDate targetDate, SortedSet<Long> buckets) {
        long[] hourly = new long[24];
        for (long bucket : buckets) {
            java.time.LocalDateTime local = Instant.ofEpochSecond(bucket * BUCKET_SECONDS)
                    .atZone(ZoneId.systemDefault())
                    .toLocalDateTime();
            if (targetDate.equals(local.toLocalDate())) {
                hourly[local.getHour()] += BUCKET_SECONDS;
            }
        }
        return hourly;
    }

    private static long[] buildSessionSeconds(SortedSet<Long> buckets) {
        List<Long> sessions = new ArrayList<>();
        Long previousBucket = null;
        long currentSeconds = 0L;

        for (long bucket : buckets) {
            if (previousBucket != null) {
                long gapSeconds = (bucket - previousBucket) * BUCKET_SECONDS;
                if (gapSeconds > SESSION_BREAK_SECONDS) {
                    if (currentSeconds > 0L) {
                        sessions.add(currentSeconds);
                    }
                    currentSeconds = 0L;
                }
            }

            currentSeconds += BUCKET_SECONDS;
            previousBucket = bucket;
        }

        if (currentSeconds > 0L) {
            sessions.add(currentSeconds);
        }

        long[] values = new long[sessions.size()];
        for (int i = 0; i < sessions.size(); i++) {
            values[i] = sessions.get(i);
        }
        return values;
    }

    private static LocalDate localDateForBucket(long bucket) {
        return Instant.ofEpochSecond(bucket * BUCKET_SECONDS)
                .atZone(ZoneId.systemDefault())
                .toLocalDate();
    }

    private static long floorBucket(long unixSeconds) {
        return unixSeconds / BUCKET_SECONDS;
    }

    private static long ceilBucket(long unixSeconds) {
        return (unixSeconds + BUCKET_SECONDS - 1L) / BUCKET_SECONDS;
    }
}
